### PR TITLE
VLCPlaybackController: Save and restore playback state with new library

### PR DESCRIPTION
**Pull request by carolanitz**
Delete old Medialibrary methods for playback
Add convenience properties on VLCMLMedia

### Checklist
- [x] I've run `bundle exec fastlane test` from the root directory to see all new and existing tests pass
- [x] I've followed the [vlc-ios code style](Docs/CodingStyle.md)
- [x] I've read the [Contribution Guidelines](https://github.com/videolan/vlc-ios#contribute)
- [x] I've updated the documentation if necessary.

### Description

I moved the saving out of the playbackcontroller because that piece of code should not be in charge  of modifying our models. This PR should also be thoroughly tested before being merged and the discussed crash should be addressed beforehand

### CODE REVIEW

File: SharedSources/MediaLibraryModel/MediaModel.swift [hunk 1]
```diff
@@ -48,16 +48,67 @@ extension VLCMLMedia {
         return ByteCountFormatter.string(fromByteCount: Int64(mainFile().size()),
                                          countStyle: .file)
     }
+}
+
+// convenience properties
+
+@objc extension VLCMLMedia {
```
Comment (carolanitz) [inline SharedSources/MediaLibraryModel/MediaModel.swift]: These properties could potentially live in the MediaLibrary

File: SharedSources/MediaLibraryService.swift [hunk 2]
```diff
@@ -284,10 +284,25 @@ extension MediaLibraryService {
...
+            assertionFailure("no mrl")
             return nil
         }
         return medialib.media(withMrl: mrl)
     }
+
+    func savePlaybackState(from player: VLCPlaybackController) {
+
+        let media = player.currentlyPlayingMedia
+        let mlMedia = fetchMedia(with: media.url.absoluteURL)!
```
Comment (carolanitz) [inline SharedSources/MediaLibraryService.swift]: for some reason this call constantly frequently crashes in Medialibrary
Details 
```
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 4.1
  * frame #0: 0x0000000105a9b0df VLC for iOS`::__cxa_throw(void *, std::type_info *, void (*)(void *)) [inlined] _dispatch_once(block=<unavailable>) block_pointer) at once.h:83:39 [opt]
    frame #1: 0x0000000105a9b0df VLC for iOS`::__cxa_throw(exception_object=0x0000600001911e80, tinfo=0x000000010b075550, dest=(libc++.1.dylib`std::__1::system_error::~system_error()))(void *)) at BITCrashCXXExceptionHandler.mm:85 [opt]
    frame #2: 0x00000001078cb432 VLC for iOS`medialibrary::utils::fs::toAbsolute(path="/Users/caro/Library/Developer/CoreSimulator/Devices/34192C85-6569-4443-951B-045D45C24E39/data/Containers/Data/Application/015AE273-A6C5-4752-9C67-F072076E268E/Documents/FMA%20-%20MultiChapter.mkv") at Directory.cpp:88:9
    frame #3: 0x000000010782f90a VLC for iOS`medialibrary::factory::FileSystemFactory::createDeviceFromMrl(this=0x0000600000c0d1b8, mrl="file:///Users/caro/Library/Developer/CoreSimulator/Devices/34192C85-6569-4443-951B-045D45C24E39/data/Containers/Data/Application/015AE273-A6C5-4752-9C67-F072076E268E/Documents/FMA%2520-%2520MultiChapter.mkv") at FileSystemFactory.cpp:81:30
    frame #4: 0x0000000107656feb VLC for iOS`medialibrary::File::fromMrl(ml=0x00007fd349e00d60, mrl="file:///Users/caro/Library/Developer/CoreSimulator/Devices/34192C85-6569-4443-951B-045D45C24E39/data/Containers/Data/Application/015AE273-A6C5-4752-9C67-F072076E268E/Documents/FMA%2520-%2520MultiChapter.mkv") at File.cpp:287:30
    frame #5: 0x00000001076f17a0 VLC for iOS`medialibrary::MediaLibrary::media(this=0x00007fd349e00d60, mrl="file:///Users/caro/Library/Developer/CoreSimulator/Devices/34192C85-6569-4443-951B-045D45C24E39/data/Containers/Data/Application/015AE273-A6C5-4752-9C67-F072076E268E/Documents/FMA%2520-%2520MultiChapter.mkv") const at MediaLibrary.cpp:376:12
    frame #6: 0x000000010758f88b VLC for iOS`::-[VLCMediaLibrary mediaWithMrl:](self=0x0000600000c18000, _cmd="mediaWithMrl:", mrl="file:///Users/caro/Library/Developer/CoreSimulator/Devices/34192C85-6569-4443-951B-045D45C24E39/data/Containers/Data/Application/015AE273-A6C5-4752-9C67-F072076E268E/Documents/FMA%2520-%2520MultiChapter.mkv") at VLCMediaLibrary.m:167:54
    frame #7: 0x00000001053940b8 VLC for iOS`MediaLibraryService.fetchMedia(mrl="file:///Users/caro/Library/Developer/CoreSimulator/Devices/34192C85-6569-4443-951B-045D45C24E39/data/Containers/Data/Application/015AE273-A6C5-4752-9C67-F072076E268E/Documents/FMA%2520-%2520MultiChapter.mkv", self=0x0000600002615140) at MediaLibraryService.swift:290:25
    frame #8: 0x00000001053943fd VLC for iOS`@objc MediaLibraryService.fetchMedia(with:) at <compiler-generated>:0
    frame #9: 0x00000001052a7dbb VLC for iOS`-[VLCMovieViewController mediaForPlayingMedia:](self=0x00007fd34a04be00, _cmd="mediaForPlayingMedia:", media=0x0000600000b754f0) at VLCMovieViewController.m:1103:12
    frame #10: 0x00000001052988a1 VLC for iOS`-[VLCPlaybackController _recoverLastPlaybackStateOfItem](self=0x0000600001d06d00, _cmd="_recoverLastPlaybackStateOfItem") at VLCPlaybackController.m:1087:25
    frame #11: 0x0000000105293740 VLC for iOS`-[VLCPlaybackController observeValueForKeyPath:ofObject:change:context:](self=0x0000600001d06d00, _cmd="observeValueForKeyPath:ofObject:change:context:", keyPath=@"remainingTime", object=0x0000600001a190e0, change=0x0000600003250f80, context=0x0000000000000000) at VLCPlaybackController.m:340:13
    frame #12: 0x000000010d8683b8 Foundation`NSKeyValueNotifyObserver + 332
    frame #13: 0x000000010d86b9dc Foundation`NSKeyValueDidChange + 489
    frame #14: 0x000000010d8679c9 Foundation`NSKeyValueDidChangeWithPerThreadPendingNotifications + 143
    frame #15: 0x0000000105ac9b35 VLC for iOS`-[VLCMediaPlayer(self=0x0000600001a190e0, _cmd=<unavailable>, newMedia=<unavailable>) mediaPlayerMediaChanged:] at VLCMediaPlayer.m:1526:9 [opt]
    frame #16: 0x0000000105abcd92 VLC for iOS`-[VLCEventManager callObjectMethodWithArgs:](self=<unavailable>, _cmd=<unavailable>, message=<unavailable>) at VLCEventManager.m:350:13 [opt]
    frame #17: 0x000000011163edb5 libdispatch.dylib`_dispatch_client_callout + 8
    frame #18: 0x000000011164ce72 libdispatch.dylib`_dispatch_async_and_wait_invoke + 175
    frame #19: 0x000000011163edb5 libdispatch.dylib`_dispatch_client_callout + 8
    frame #20: 0x000000011164c080 libdispatch.dylib`_dispatch_main_queue_callback_4CF + 1540
    frame #21: 0x000000010fe3e8a9 CoreFoundation`__CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
    frame #22: 0x000000010fe38f56 CoreFoundation`__CFRunLoopRun + 2310
    frame #23: 0x000000010fe38302 CoreFoundation`CFRunLoopRunSpecific + 626
    frame #24: 0x0000000112cd32fe GraphicsServices`GSEventRunModal + 65
    frame #25: 0x000000011c5dcba2 UIKitCore`UIApplicationMain + 140
    frame #26: 0x00000001052e6dc0 VLC for iOS`main(argc=1, argv=0x00007ffeea98df98) at main.m:20:16
    frame #27: 0x00000001116b3541 libdyld.dylib`start + 1
    frame #28: 0x00000001116b3541 libdyld.dylib`start + 1
```

File: SharedSources/MediaLibraryService.swift [hunk 3]
```diff
@@ -284,10 +284,25 @@ extension MediaLibraryService {
...
+
+        mlMedia.isNew = false
+        mlMedia.progress = player.playbackPosition
+        mlMedia.audioTrackIndex = player.indexOfCurrentAudioTrack
+        mlMedia.subtitleTrackIndex = player.indexOfCurrentSubtitleTrack
+        mlMedia.chapterIndex = player.indexOfCurrentChapter
+        mlMedia.titleIndex = player.indexOfCurrentTitle
+        //create a new thumbnail
```
Comment (carolanitz) [inline SharedSources/MediaLibraryService.swift]: the old code triggered a thumbnail update with the current position not sure if this is doable with the new medialibrary

File: SharedSources/VLCMetadata.m [hunk 4]
```diff
@@ -69,7 +68,7 @@ - (void)updateMetadataFromMediaPlayerForiOS:(VLCMediaPlayer *)mediaPlayer
     [self checkIsAudioOnly:mediaPlayer];
 
     if (self.isAudioOnly) {
-        self.artworkImage = [VLCThumbnailsCache thumbnailForManagedObject:item];
```
Comment (carolanitz) [inline SharedSources/VLCMetadata.m]: This method was removed and we should add a ticket for adapting this part of the code to the new medialibrary

File: Sources/VLCAppDelegate.m [hunk 5]
```diff
@@ -190,7 +190,7 @@ - (BOOL)application:(UIApplication *)application
             return NO;
         }
         [self validatePasscodeIfNeededWithCompletion:^{
-            [[VLCPlaybackController sharedInstance] openMediaLibraryObject:managedObject];
+           // [[VLCPlaybackController sharedInstance] openMediaLibraryObject:managedObject];
```
Comment (carolanitz) [inline Sources/VLCAppDelegate.m]: need to add ticket to update this part of the code for useractivity

File: Sources/VLCPlaybackController.m [hunk 6]
```diff
@@ -322,94 +322,13 @@ - (void)stopPlayback
...
 {
-    MLFile *item = [MLFile fileForURL:_mediaPlayer.media.url].firstObject;
+    VLCMLMedia *media = [_delegate mediaForPlayingMedia:_mediaPlayer.media];
 
-    if (item) {
-        _mediaPlayer.currentAudioTrackIndex = item.lastAudioTrack.intValue;
-        _mediaPlayer.currentVideoSubTitleIndex = item.lastSubtitleTrack.intValue;
+    if (media) {
+        _mediaPlayer.currentAudioTrackIndex = (int) media.audioTrackIndex;
+        _mediaPlayer.currentVideoSubTitleIndex = (int) media.subtitleTrackIndex + 1;
```
Comment (carolanitz) [inline Sources/VLCPlaybackController.m]: for some reason directly storing and reaplying currentVideoSubTitleIndex is off by one

File: Sources/VLCPlaybackController.m [hunk 7]
```diff
@@ -1165,46 +1077,40 @@ - (void)setNeedsMetadataUpdate
...
-                                                     errorMessage:[NSString stringWithFormat:NSLocalizedString(@"CONTINUE_PLAYBACK_LONG", nil), item.title]
-                                                   viewController:presentingVC
-                                                    buttonsAction:buttonsAction];
+- (void)_recoverLastPlaybackState
+{
+    VLCMLMedia *media = [_delegate mediaForPlayingMedia:_mediaPlayer.media];
+    if (!media) return;
+
+    CGFloat lastPosition = media.progress;
+    if (_mediaPlayer.position < lastPosition) {
```
Comment (carolanitz) [inline Sources/VLCPlaybackController.m]: note that I removed the .95 percent check knowingly, when you have tv shows it tended to not restart even though you weren't finished

File: SharedSources/MediaLibraryService.swift (hunk 2 above)
Comment (chouquette) [reply SharedSources/MediaLibraryService.swift]: `FMA%20-%20MultiChapter.mkv`
Is the %20 actually part of the file name? Or is that a left over from a wrong/missing mrl decode?

Comment (chouquette) [reply SharedSources/MediaLibraryService.swift]: Although we changed the mrl encode/decode recently, so you might have an unmigrated database, causing this crash in the end. Could you try flushing your database?

Comment (carolanitz) [reply SharedSources/MediaLibraryService.swift]: the %20 is not part of the file name it's a space

Comment (chouquette) [reply SharedSources/MediaLibraryService.swift]: This smells a lot like a double encoding happened somewhere. It should be `%20` all the way, but it starts with `%2520`, which is a `%` encoded, followed by a 20. So the percent gets decoded, but the 20 is now just a plain 20.
Where is the mrl you're passing to the medialibrary coming from?

Comment (carolanitz) [reply SharedSources/MediaLibraryService.swift]: it directly pipes the url from a VLCMedia in there

Comment (chouquette) [reply SharedSources/MediaLibraryService.swift]: Please bear with my lack of iOS codebase knowledge, but ultimately, where does this MRL comes from?
TL;DR: `%2520` is very (very very) likely an encoding screw up. The question is to know where it happened.
In any case, the medialibrary is "expected" to throw an exception here, since it tries to locate a file that doesn't exist. I need to add the appropriate catch though (https://code.videolan.org/videolan/medialibrary/issues/80)

Comment (carolanitz) [reply SharedSources/MediaLibraryService.swift]: :) sorry I will double check the internals of VLCKit to see if I find the double encoding!

File: SharedSources/MediaLibraryModel/MediaModel.swift (hunk 1 above)
Comment (Mikanbu) [reply SharedSources/MediaLibraryModel/MediaModel.swift]: Indeed, I'm not sure the app-side is the best place for all the properties.

File: SharedSources/MediaLibraryService.swift (hunk 3 above)
Comment (Mikanbu) [reply SharedSources/MediaLibraryService.swift]: Indeed, related ticket [here](https://code.videolan.org/videolan/medialibrary/issues/82).

File: SharedSources/MediaLibraryService.swift (hunk 2 above)
Comment (carolanitz) [reply SharedSources/MediaLibraryService.swift]: turns out this unveiled a longstanding bug in vlckit. Will be merged and fixed with the next VLCKit release

**Comment by carolanitz** [comment]
ticket that needs to be filed after this is merged. The state is saved and correctly resumed but the ui isn't updated after the player is dismissed or while the media plays

File: SharedSources/MediaLibraryService.swift [hunk 8]
```diff
@@ -288,10 +288,28 @@ extension MediaLibraryService {
 
     func fetchMedia(with mrl: URL?) -> VLCMLMedia? {
         guard let mrl = mrl  else {
+            assertionFailure("no mrl")
```
Comment (Mikanbu) [inline SharedSources/MediaLibraryService.swift]: I believe it could be better to have a better string for the assertion see the other assertions on in the service.

Example: `assertionFailure("MediaLibraryService: Medialibrary already initialized.")`

File: Sources/MiniPlayer/AudioMiniPlayer.swift [hunk 9]
```diff
@@ -12,10 +12,12 @@
 
 @objc(VLCAudioMiniPlayer)
 class AudioMiniPlayer: UIView, MiniPlayer {
+
```
Comment (Mikanbu) [inline Sources/MiniPlayer/AudioMiniPlayer.swift]: Nit-pick: unnecessary cosmetic changes.

File: Sources/MiniPlayer/MiniPlayer.swift [hunk 10]
```diff
@@ -14,4 +14,5 @@
 protocol MiniPlayer: VLCMiniPlaybackViewInterface, VLCPlaybackControllerDelegate {
     var contentHeight: Float { get }
     func updatePlayPauseButton()
+    var mediaService: MediaLibraryService? { get set }
```
Comment (Mikanbu) [inline Sources/MiniPlayer/MiniPlayer.swift]: I think it would be better if it was directly passed to the specific player than adding it to the protocol.

File: Sources/VLCLibrarySearchDisplayDataSource.m [hunk 11]
```diff
@@ -11,6 +11,7 @@
  *****************************************************************************/
 
 #import "VLCLibrarySearchDisplayDataSource.h"
+#import <MediaLibraryKit/MediaLibraryKit.h>
```
Comment (Mikanbu) [inline Sources/VLCLibrarySearchDisplayDataSource.m]: Is this needed for restoration?

**Comment by carolanitz** [comment]
merged with 62c67bed32185964cecfb3d4db481448606794ba
